### PR TITLE
update mpas-source: add GOTM vertical mixing library

### PR DESCRIPTION
**Pull request by mark-petersen**
This PR adds the General Ocean Turbulence Model (GOTM) in MPAS-Ocean as an additional option for the vertical turbulence closure, especially for coastal simulations in which the present implementation of KPP in CVMix does not work well. GOTM is a turbulence closure library that includes a set of two equation models such as k-epsilon and the generic length scale model (GLS; Umlauf and Burchard, 2003). See https://github.com/MPAS-Dev/MPAS-Model/pull/704

GOTM is default off, but is compiled by E3SM.

[BFB]
[NML]

### CODE REVIEW

**Comment by mark-petersen** [comment]
This shows the changes in #4219. Will rebase after that is merged.

**Comment by mark-petersen** [comment]
Rebased. This has PR adds seven namelist flags. @jonbob could you run the namelist scripts? Thanks.

**Comment by mattdturner** [comment]
## Testing

The following tests pass
### Compy (PGI + Intel)
```
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.compy_intel
PET_Ln3.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi
PET_Ln9.T62_oQU240.GMPAS-IAF.compy_pgi
SMS.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.compy_pgi
SMS.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.compy_intel
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_intel
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi
PEM_Ln3.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_intel
PEM_Ln9.T62_oQU240.GMPAS-IAF.compy_intel
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.compy_pgi
PEM_Ln3.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi
PEM_Ln9.T62_oQU240.GMPAS-IAF.compy_pgi
PET_Ln3.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_intel
PET_Ln9.T62_oQU240.GMPAS-IAF.compy_intel
SMS_D.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.compy_pgi
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_intel
PET_Ln3_D.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.compy_pgi
PEM_Ln9_D.T62_oQU240.GMPAS-IAF.compy_pgi
PEM_Ln9_D.T62_oQU240.GMPAS-IAF.compy_intel
```

### Anvil (GNU + Intel)
```
SMS.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.anvil_gnu
SMS.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.anvil_intel
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_gnu
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_intel
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.anvil_gnu
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.anvil_intel
PEM_Ln9.T62_oQU240.GMPAS-IAF.anvil_gnu
PEM_Ln3.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_gnu
PEM_Ln9.T62_oQU240.GMPAS-IAF.anvil_intel
PEM_Ln3.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_intel
PET_Ln3.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_intel
PET_Ln9.T62_oQU240.GMPAS-IAF.anvil_intel
PET_Ln3.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_gnu
PET_Ln9.T62_oQU240.GMPAS-IAF.anvil_gnu
SMS_D.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.anvil_intel
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_intel
SMS_D.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.anvil_gnu
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_gnu
PET_Ln3_D.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.anvil_gnu
PEM_Ln9_D.T62_oQU240.GMPAS-IAF.anvil_intel
PEM_Ln9_D.T62_oQU240.GMPAS-IAF.anvil_gnu
```

### BFB Tests
```
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi.C
  - PASS SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi BASELINE /compyfs/e3sm_baselines/pgi/master: 
PEM_Ln9.ne30_oECv3.A_WCYCL1850S.chrysalis_intel.C
  - PASS PEM_Ln9.ne30_oECv3.A_WCYCL1850S.chrysalis_intel BASELINE /lcrc/group/e3sm/baselines/chrys/intel
/master:
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.chrysalis_intel.C
  - PASS SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.chrysalis_intel BASELINE /lcrc/group/e3sm/baselines/chrys/
intel/master:
```

**Comment by rljacob** [comment]
Why does the gotm submodule point to a personal fork https://github.com/qingli411/code instead of the main repo https://github.com/gotm-model/code ?

**Comment by jonbob** [comment]
I will run the scripts to update the bld files later today -- it's a good chance to test out the new version that @akturner and @xylar have worked on. But we will have a WC MPAS PR that probably needs to supersede this one

**Comment by mark-petersen** [comment]
@rljacob, the `spacing` Fortran intrinsic is not supported by PGI on initialization and causes a compile error on this line of the PGI library:
```
real(rk),parameter :: default_minimum = default_fill_value + spacing(default_fill_value)
```
That `default_minimum` is used in object definitions for creating netcdf files. We could only fix it by removing the `spacing` command. The personal branch is identical to gotm except for this line. See [discussion here](https://github.com/MPAS-Dev/MPAS-Model/pull/704#issuecomment-801608459) for our fix and the output of differences.

We could put a fork of gotm on `E3SM-Project`. Would that be preferable? We can also pursue asking `gotm` to change that line, but I don't know if they will be receptive or how long it would take.

**Comment by rljacob** [comment]
Yes it would be preferable to point to fork in E3SM-Projects.  Here it is:  https://github.com/E3SM-Project/GOTMcode

**Comment by mark-petersen** [comment]
Perfect. @rljacob can you give me push permission to https://github.com/E3SM-Project/GOTMcode? I need to set it to https://github.com/qingli411/code/commit/48104b94458398ba668fc381837202dfbe580dab, which is V5.3 plus one commit for the PGI error.

**Comment by rljacob** [comment]
Done.  If you think you'll be making more changes to GOTM, you may want to create an "e3sm-develop" branch, commit changes there and point the submodule to that.  Then you could keep master of the fork as a copy of the upstream master.

**Comment by mark-petersen** [comment]
@rljacob I followed your suggestion and added a new branch `e3sm/develop` so we can point to that one, and `master` can just follow the GOTM repo.

**Comment by jonbob** [comment]
bld files are updated using the automated scripts

**Comment by mark-petersen** [comment]
From [previous tests](https://github.com/MPAS-Dev/MPAS-Model/pull/704#issuecomment-794440199), 
```
PASS:
SMS.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.anvil_intel
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.anvil_intel
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_intel
PEM_Ln3.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_intel
PEM_Ln9.T62_oQU240.GMPAS-IAF.anvil_intel
PET_Ln3.T62_oQU120_ais20.MPAS_LISIO_TEST.anvil_intel
PET_Ln9.T62_oQU240.GMPAS-IAF.anvil_intel
PET_Ln3_D.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.anvil_intel
PEM_Ln9_D.T62_oQU240.GMPAS-IAF.anvil_intel
SMS.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-haswell_gnu
SMS.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-haswell_intel
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-haswell_intel
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-haswell_gnu
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.cori-haswell_intel
PEM_Ln9.T62_oQU240.GMPAS-IAF.cori-haswell_gnu
PEM_Ln9.T62_oQU240.GMPAS-IAF.cori-haswell_intel
PET_Ln9_P8192.T62_oRRS18to6v3.GMPAS-IAF.cori-knl_gnu
PET_Ln9_P8192.T62_oRRS18to6v3.GMPAS-IAF.cori-knl_intel
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_gnu
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.cori-knl_intel
PET_Ln9.T62_oQU240.GMPAS-IAF.cori-knl_intel
```

**Comment by jonbob** [comment]
test merge passes:
ERS.T62_EC30to60E2r2.GMPAS-IAF.anvil_intel

**Review by jonbob — APPROVED** [review]
Approved based on visual inspection, developer testing, and sanity tests

**Comment by mattdturner** [comment]
I realize this was already merged to next, but I was asked to test PGI to confirm that nothing weird happened when switching the GOTM repo.

```
PASS:
SMS.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.compy_pgi
SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi
PEM_Ln9.T62_oQU240.GMPAS-IAF.compy_pgi
PEM_Ln3.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi
PET_Ln3.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.compy_pgi
PET_Ln9.T62_oQU240.GMPAS-IAF.compy_pgi
PET_Ln3.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi
SMS_D.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.compy_pgi
SMS_D.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi
PET_Ln3_D.T62_oEC60to30v3wLI.GMPAS-DIB-IAF-ISMF.compy_pgi
PEM_Ln9_D.T62_oQU240.GMPAS-IAF.compy_pgi

--
FAIL:
FAIL SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi NLCOMP
  - PASS SMS.T62_oQU120_ais20.MPAS_LISIO_TEST.compy_pgi BASELINE /compyfs/e3sm_baselines/pgi
/master:
```

The FAIL is just a namelist fail (expected)

**Comment by jonbob** [comment]
merged to master and expected NML DIFFs blessed

**Pull request merged**
